### PR TITLE
Try to fix the setup wizard trouble on JN

### PR DESCRIPTION
**Pull request by peterfabian**
# Background
OBW was skipped on JN, even though it was working fine on local machine. The problem, as I discovered, was caused by actions being trigerred too quickly/often on JN, which cleared up the transient instantly and (sometimes) saved incorrect values of notices, i.e. changed values of variables affecting whether merchants is redirected to OBW or not. 

## Scenario 1

1. Click `Activate plugin`, which runs /wp-admin/plugins.php?action=activate&plugin=woocommerce%2Fwoocommerce.php&_wpnonce=57c117faca
2. during the run (here https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-install.php#L350)
4. run WC_Admin_Notices::add_notice('install') (this however, only saves to memory)
3. set transient _wc_activation_redirect: 1 (this saves to db), 
3. in the bg, AS is trigerred: /wp-admin/admin-ajax.php?action=as_async_request_queue_runner&nonce=7388fe17ef
4. This would trigger `\WC_Admin::admin_redirects` which would read `WC_Admin_Notices` from the database, but as nothing have been saved yet, transient `_wc_activation_redirect` is deleted here: https://github.com/woocommerce/woocommerce/blob/master/includes/admin/class-wc-admin.php#L158
5. By the time regular request [gets to test whether the merchant should be redirected](https://github.com/woocommerce/woocommerce/blob/master/includes/admin/class-wc-admin.php#L147) to OBW, the transient is gone, so no OBW for the merchant.

**Solution**: don't run `admin_redirects` during AS requests. It's not needed.

## Scenario 2 
Even if the transient clearing is removed (which causes the OBW to be displayed, yay), the notice to display the notification to continue with OBW if merchant quits it initially can be lost as well:

1. Click `Activate plugin`, which runs /wp-admin/plugins.php?action=activate&plugin=woocommerce%2Fwoocommerce.php&_wpnonce=57c117faca
2. during the run (here https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-install.php#L350)
4. run WC_Admin_Notices::add_notice('install') (this however, only saves to memory)
3. set transient _wc_activation_redirect: 1 (this saves to db), 
4. in the context of AS, read notices from db (aka nothing)
5. in the context of normal request, save notices to db on shutdown
6. get redirected, another regular request, read the notices from the db
7. redirect to OBW, yay!, but I don't want it now, going back to dashboard
8. AS request from 5 is finishing, saves 0 notices to db
9. No more notices to finish OBW in the dashboard.

**Solution**: don't save notices, if we're in installation phase & running in context of AS.

# Generalization
Basically, any time we load something on init and save it on shutdown, we need to expect AS to run any time in between… so e.g.
1. AS init: read_x; sees x = null;
2. regular request init: do stuff that results in set_x = 3;
3. regular request shutdown: store_x = 3 to db;
4. AS shutdown: store_x = null;

or 1, 2, 4, 3 (between 4 and 3 weird stuff might happen)
or any combo of these…

I’m not sure what’s the best way to prevent this from happening… I’m adding some conditionals to the code (i.e. if you’re running from AS context, don’t save the values), but that is probably not always viable

@rrennick advised there's a $context parameter when action is being run for async processing. 

However, I'm afraid it does not matter if it’s an async request or if it runs via cron, in either of those cases, we need to avoid the side effects from AS. 

## How to test this PR

1. Check out the branch 
2. Build a package, but you need to apply the fix from https://github.com/woocommerce/action-scheduler/pull/450
3. Install on JN and atomic, ideally multiple times, it should always show OBW, 
4. if OBW is cancelled, it should show a notification to run it.

### CODE REVIEW

**Comment by rrennick** [comment]
@peterfabian I'm wondering whether disabling the async runner until the onboarding flow is completed or dismissed might be a better option. If the site is in the onboarding flow, it's unlikely that they have a large queue of pending actions. The filter for that is https://github.com/woocommerce/action-scheduler/blob/master/classes/ActionScheduler_AsyncRequest_QueueRunner.php#L87

If that would be the better route then if it's the new onboarding experience would it be better to do that in WC Admin?

File: bin/package-update.sh [hunk 1]
```diff
@@ -39,3 +39,10 @@ find ./packages/woocommerce-admin -iname '*.js' -exec sed -i.bak -e "s/, 'woocom
 # Cleanup backup files
 find ./packages -name "*.bak" -type f -delete
 output 2 "Done!"
+
+# Apply patches
+output 2 "Applying patch #450 to Action Schduler"
+cd packages/action-scheduler
+curl -O https://patch-diff.githubusercontent.com/raw/woocommerce/action-scheduler/pull/450.patch
+patch -p1 < 450.patch
```
Comment (vedanshujain) [inline bin/package-update.sh]: [not a blocker] should it be ` patch -p1 < 450.patch && output 2 "Done!" `?

File: includes/admin/class-wc-admin-notices.php [hunk 2]
```diff
@@ -49,7 +49,9 @@ public static function init() {
 		add_action( 'switch_theme', array( __CLASS__, 'reset_admin_notices' ) );
 		add_action( 'woocommerce_installed', array( __CLASS__, 'reset_admin_notices' ) );
 		add_action( 'wp_loaded', array( __CLASS__, 'hide_notices' ) );
-		add_action( 'shutdown', array( __CLASS__, 'store_notices' ) );
+		if ( ! WC_Install::is_new_install() || ! wc_is_running_from_async_action_scheduler() ) {
```
Comment (vedanshujain) [inline includes/admin/class-wc-admin-notices.php]: Probably add a comment here about why because it won't be super clear otherwise

File: includes/wc-core-functions.php [hunk 3]
```diff
@@ -2273,3 +2273,13 @@ function wc_load_cart() {
...
 }
+
+/**
+ * Test whether the context of execution comes from async action scheduler.
+ *
+ * @since 4.0.0
+ * @return bool
+ */
+function wc_is_running_from_async_action_scheduler() {
+	return isset( $_REQUEST['action'] ) && 'as_async_request_queue_runner' === $_REQUEST['action'];
```
Comment (vedanshujain) [inline includes/wc-core-functions.php]: I wonder if this is the best way to check whether AS is running or not? but maybe we can fix this in a later PR, cc @rrennick

**Review by vedanshujain — APPROVED** [review]
Left minor comments, other than LGTM!

File: includes/wc-core-functions.php (hunk 3 above)
Comment (peterfabian) [reply includes/wc-core-functions.php]: 👍 Yup, I think this needs a revisit. Maybe even to check for cron actions? More investigation is needed.

File: bin/package-update.sh (hunk 1 above)
Comment (peterfabian) [reply bin/package-update.sh]: Yeah... maybe not. Due to our dependency, I _think_ we need to run the following steps to get the package:

```
npm install
# this includes wp cli i18n package 
composer install
# this uses wp cli i18n package 
npm run build 
# to get rid of all packages we don't need
composer install --no-dev
```

Second patch fails (already applied) and I don't think we need to fix it now, as we should get rid of extra patching anyway.

File: includes/admin/class-wc-admin-notices.php (hunk 2 above)
Comment (peterfabian) [reply includes/admin/class-wc-admin-notices.php]: Good idea, added.

**Comment by peterfabian** [comment]
We'll need to revisit this and find a better solution, but for now, it seems to test fine.

**Comment by rrennick** [comment]
@peterfabian That sounds good. It's possible that other plugins could have a similar flow to AS so a long term solution shouldn't be specific to AS.

**Comment by oskosk** [comment]
@peterfabian this may have been my fault. There was a piece of code in JN sites which explicitly deleted the transient `_wc_activation_redirect`. JN sites carry an unlisted plugin which is in charge of the auto login. This plugi is called companion. I just fixed this (for every new site launched from now on) in this commit https://github.com/Automattic/companion/commit/ddc9212a28cec72067ff35c8580c2003c8f4e20f. I'm sorry if this line was guilty and caused confusion.
